crates/ty_python_semantic/resources/mdtest/generics/pep695/variance.md

```diff
@@ -380,4 +336,296 @@ static_assert(not is_equivalent_to(D[Any], C[Any]))
+    def f(self) -> "D[X]":
+        return D()
+
+    def g(self, x: X) -> None: ...
+
+class D[Y]:
+    def h(self) -> C[Y]:
+        return C()
+```
+
+`X` is contravariant in both `C` and `D`:
+
+- `C` has two occurrences of `X`
+    - `X` occurs in the return type of `f` as `D[X]` (`X` is substituted in for `Y`)
+        - `D` has one occurrence of `Y`
+            - `Y` occurs in the return type of `h` as `C[Y]`
+    - `X` occurs contravariantly as a parameter in `g`
+
+We want the _greatest_ inferrable variance, so we start by assuming `X` occurs bivariantly in `C`.
+Therefore, `X` has one bivariant occurrence, and one contravariant occurrence, which resolves to
+contravariant.
+
+Recall that `Y`'s only occurrence in `D` is covariantly in `C[Y]` (`Y` subbing in for `X`). We just
+stated that `X` is contravariant in `C`, so we conclude that `Y` is contravariant in `D`.
+
+```py
+static_assert(not is_subtype_of(C[B], C[A]))
+static_assert(is_subtype_of(C[A], C[B]))
+static_assert(not is_subtype_of(C[A], C[Any]))
+static_assert(not is_subtype_of(C[B], C[Any]))
+static_assert(not is_subtype_of(C[Any], C[A]))
+static_assert(not is_subtype_of(C[Any], C[B]))
+
+static_assert(not is_subtype_of(D[B], D[A]))
+static_assert(is_subtype_of(D[A], D[B]))
+static_assert(not is_subtype_of(D[A], D[Any]))
+static_assert(not is_subtype_of(D[B], D[Any]))
+static_assert(not is_subtype_of(D[Any], D[A]))
+static_assert(not is_subtype_of(D[Any], D[B]))
+```
+
+## Class Attributes
+
+### Mutable Attributes
+
+Normal attribtues are mutable, and so are invariant (see [inv]).
+
+```py
+from ty_extensions import is_subtype_of, static_assert
+
+class A: ...
+class B(A): ...
+
+class C[T]:
+    x: T
+
+static_assert(not is_subtype_of(C[B], C[A]))
+static_assert(not is_subtype_of(C[A], C[B]))
+```
+
+One might think that normal attributes are
+
+### Immutable Attributes
+
+Immutable attributes can't be written to, and so don't have the same problem as mutable ones.
+
+```py
+from dataclasses import dataclass
+from typing import Final, NamedTuple
+from ty_extensions import is_subtype_of, static_assert
+
+class A: ...
+class B(A): ...
+
+class C[T]:
+    x: Final[T]
+
+@dataclass(frozen=True)
+class D[U]:
+    y: U
+
+class E[V](NamedTuple):
+    z: V
+
+static_assert(is_subtype_of(C[B], C[A]))
+static_assert(not is_subtype_of(C[A], C[B]))
+
+static_assert(is_subtype_of(D[B], D[A]))
+static_assert(not is_subtype_of(D[A], D[B]))
+
+static_assert(is_subtype_of(E[B], E[A]))
+static_assert(not is_subtype_of(E[A], E[B]))
+```
+
+### Properties
+
+Properties are covariant if they are get-only and invariant if they are getset
+
+```py
+class A: ...
+class B(A): ...
+
+class C[T]:
+    @property
+    def x(self) -> T | None:
+        return None
+
+class D[U]:
+    @property
+    def y(self) -> U | None:
+        return None
+
+    @y.setter
+    def value(self, value: U): ...
+```
+
+while accessing `D().x` will give you an `int`, the descriptor methods say this should be
```

Comment:
D doesn't have an `x` attribute

Reply:
I think this was me trying to communicate a point you made in discord---we want to access members without going through the descriptor protocol.

The int thing is a goof on my part. I was trying to come up with an example where the return type of the getter would be deceiving versus that of the descriptor variance-wise, and was trying to imagine a descriptor whose getter would return `int` even though it stored a `T`, but I couldn't really do it, and never cleaned this up--sorry!